isic/ingest/services/publish/__init__.py

```diff
@@ -59,49 +61,56 @@ def cohort_publish_initialize(
             public=public,
         )
         publish_request.accessions.set(cohort.accessions.publishable())
+        publish_request.collections.set(collections)
         # ensure that the magic collection is also added to the publish request
-        publish_request.collections.set(
-            collections.union(Collection.objects.filter(id=cohort.collection.id))
-        )
+        publish_request.collections.add(cohort.collection)
 
     publish_cohort_task.delay_on_commit(publish_request.pk)
 
 
 def cohort_publish(*, publish_request: PublishRequest) -> None:
-    # this creates a transaction
-    with lock_table_for_writes(IsicId), transaction.atomic():
-        for accession in publish_request.accessions.iterator():
-            if accession.attribution == "":
-                accession.attribution = accession.cohort.default_attribution
-                accession.save(update_fields=["attribution"])
-
-            image_create(
-                creator=publish_request.creator,
-                accession=accession,
-                public=False,
-            )
-        if not publish_request.public:
-            for collection in publish_request.collections.all():
-                collection_add_images(
-                    collection=collection,
-                    qs=Image.objects.filter(accession__in=publish_request.accessions.all()),
-                    ignore_lock=True,
-                )
-
-    if publish_request.public:
-        with transaction.atomic():
-            unembargo_images(
-                qs=Image.objects.filter(accession__in=publish_request.accessions.all())
-            )
+    from isic.ingest.tasks import publish_accession_task
+
+    additional_collection_ids = list(publish_request.collections.values_list("id", flat=True))
+
+    for accession_pk in throttled_iterator(
+        publish_request.accessions.values_list("pk", flat=True).iterator()
+    ):
+        publish_accession_task.delay_on_commit(
+            accession_pk=accession_pk,
+            public=publish_request.public,
+            publisher_pk=publish_request.creator.pk,
+            additional_collection_ids=additional_collection_ids,
+        )
+
+
+def accession_publish(
+    *,
+    accession: Accession,
+    public: bool,
+    publisher: User,
+    additional_collection_ids: Iterable[int] | None = None,
+):
+    with lock_table_for_writes(IsicId):  # this also opens a transaction
```

Comment:
Code comment here that wrapping the unembargo and collection add into the transaction makes this atomic and easier to completely retry if any part fails.